client/core/core.go

```diff
@@ -5118,6 +5111,48 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
 	return nil
 }
 
+// acceptCsum will record the commitment checksum so we can verify that the
+// subsequent match_proof with this order has the same checksum. If it does not,
+// the server may have used the knowledge of this preimage we are sending them
+// now to alter the epoch shuffle.
+func acceptCsum(dc *dexConnection, reqID uint64, tracker *trackedTrade, isCancel bool, commitChecksum dex.Bytes) error {
+	var respondWithErr bool
+
+	// Allow to initialize csum only once per order to prevent possible
+	// malicious behavior.
+	tracker.mtx.Lock()
+	if isCancel {
+		if tracker.cancel.csum != nil {
+			respondWithErr = true
+		} else {
+			tracker.cancel.csum = commitChecksum
+		}
+	} else {
+		if tracker.csum != nil {
+			respondWithErr = true
+		} else {
+			tracker.csum = commitChecksum
+		}
+	}
+	tracker.mtx.Unlock()
+
+	if respondWithErr {
+		csumErr := errors.New("csum is already initialized")
+		resp, err := msgjson.NewResponse(reqID, nil,
+			msgjson.NewError(msgjson.InvalidRequestError, csumErr.Error()))
+		if err != nil {
+			return fmt.Errorf("preimage response encoding error: %w", err)
+		}
+		err = dc.Send(resp)
+		if err != nil {
+			return fmt.Errorf("preimage send error: %w", err)
+		}
+		return csumErr
+	}
```

Comment:
> What's the reasoning for any response at all?

The reasoning is - having some response instead of no response or unknown response would make testing/debugging any related issues easier for ourselves.

> It would seem to be a courtesy to a server that is misbehaving.

Since this code is open source I see no value in security through obscurity here.

> That's the present approach.

Although I wasn't able to implement it properly on my first try, I'll prefer the approach 4) "returning a meaningful error", 
if not then returning no response and just timing out seems like the simplest to implement (and cleanest to understand) option among those you've listed above, so I would implement it instead.

Reply:
> Since this code is open source I see no value in security through obscurity here

I'm not implying it's a security benefit to not respond, but responding does allow the server issuing the request to move on with the epoch matching sooner instead of waiting for a timeout, which is something of an undeserved gift to a server that is misbehaving.  I'm OK with a response at this point because it does assist debugging and minimize disruption due to just bugs, but I had envisioned just not responding...  Just wanted to point out that if we get to a point where it's clear that if this does happen that it is genuinely malicious, not a bug, then client probably should give server the cold shoulder.

> Although I wasn't able to implement it properly on my first try, I'll prefer the approach 4) "returning a meaningful error",
if not then returning no response and just timing out seems like the simplest to implement (and cleanest to understand) option among those you've listed above, so I would go for it instead

You're pretty close to this.  As I suggested in the main comment for this review, the `handlePreimageResp` method can get modified to check `if resp.Error != nil {` before `json.Unmarshal`, which would probably be the minimal code change.  This is similar to `msg.UnmarshalResult`, but I don't think we always want to `m.repsondError` (see below).

Note that in `handlePreimageResp`, all the `m.respondError` calls are kinda dumb because client won't find a response handler since this isn't a client-originating request, but it will at least log something at the client.  But if it detects a non-nil `resp.Error` field between `msg.Response` and `json.Unmarshal`, that condition should omit the `m.respondError` because the client needs no further response.

i.e.

```diff
diff --git a/server/market/market.go b/server/market/market.go
index 31118188..74d62ae5 100644
--- a/server/market/market.go
+++ b/server/market/market.go
@@ -1533,6 +1533,11 @@ func (m *Market) handlePreimageResp(msg *msgjson.Message, reqData *piData) {
                        "error parsing preimage notification response")
                return
        }
+       if resp.Error != nil {
+               log.Warnf("Client failed to handle preimage request: %v", resp.Error)
+               sendPI(nil) // thanks for the response, client, we can move on now
+               return
+       }
        err = json.Unmarshal(resp.Result, &piResp)
        if err != nil {
                sendPI(nil)

```

Reply:
> if we get to a point where it's clear that if this does happen that it is genuinely malicious, not a bug, then client probably should give server the cold shoulder.

Good point, although I don't have any ideas at the moment on how to reliably differentiate between a bug and malicious behavior.

Otherwise very thorough description for what is going on, thanks for that! I've pushed these changes (added a test for these as well).